### PR TITLE
Bug - shadow variable clash with destructured props

### PR DESCRIPTION
**Pull request by harshmalde**
Issue Description:
When shadow variable name and deconstructed `props` property are same, babel is unable to differentiate between them (since both the property will be of type `Identifier` and you can have a look at the storybook that has been created with this PR).

Implementation:
This pull request helps to reconstruct the destructed props in the Arrow function.

Bug:
[APP-286](https://product-fabric.atlassian.net/browse/APP-286)

### CODE REVIEW

File: stories/css-shadow-variable-clash.tsx [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  title: 'css prop/variable clash',
+};
+
+const isPrimary = true;
+const Component1 = styled.p`
+  color: ${({ isPrimary }) => (isPrimary ? 'green' : 'red')};
+`;
+
+export const UsingComponent1 = (): JSX.Element => {
+  return <Component1 isPrimary={false}>Component1</Component1>;
```
Comment (pancaspe87) [inline stories/css-shadow-variable-clash.tsx:13]: Can you
- use a more descriptive component name, eg ComponentWithProps
- add a more descriptive text, eg 'This should be red'

File: stories/css-shadow-variable-clash.tsx [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+};
```
Comment (pancaspe87) [inline stories/css-shadow-variable-clash.tsx:14]: Can you add 1 more example without props, eg
```
export const ComponentWithoutProps = (): JSX.Element => {
  return <Component> This should be red </Component>;
};
```

File: stories/css-shadow-variable-clash.tsx [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { styled } from '@compiled/react';
+
+export default {
+  title: 'css prop/variable clash',
+};
+
+const isPrimary = true;
```
Comment (pancaspe87) [inline stories/css-shadow-variable-clash.tsx:7]: Maybe add a console.log to use the variable

File: stories/css-shadow-variable-clash.tsx [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+const Component1 = styled.p`
```
Comment (pancaspe87) [inline stories/css-shadow-variable-clash.tsx:8]: nit
I think this could be named just `Component`

**Comment by pancaspe87** [comment]
@harshmalde run `yarn changeset` and follow the prompt so that you can describe the type of change

File: packages/babel-plugin/src/utils/css-builders.ts [hunk 5]
```diff
@@ -619,6 +619,16 @@ const extractObjectExpression = (node: t.ObjectExpression, meta: Metadata): CSSO
   return { css: mergeSubsequentUnconditionalCssItems(css), variables };
 };
 
+const getMemberExpressionForOwnPath = (
```
[the remark this answers is not in the document]
Comment (harshmalde) [reply packages/babel-plugin/src/utils/css-builders.ts:629]: I have added the comments

File: packages/babel-plugin/src/utils/css-builders.ts [hunk 6]
```diff
@@ -633,6 +643,31 @@ const extractTemplateLiteral = (node: t.TemplateLiteral, meta: Metadata): CSSOut
...
 
+    // Convert destructed props into structed props
+    // Babel was not able to differentiate between destrcuted prop and global variable if they have same name
+    if (
+      t.isArrowFunctionExpression(nodeExpression) &&
+      t.isObjectPattern(nodeExpression.params[0])
+    ) {
+      const properties = nodeExpression.params[0].properties as t.ObjectProperty[];
+      const parameters = properties.map((po) => {
+        if (!t.isObjectProperty(po)) {
```
[the remark this answers is not in the document]
Comment (harshmalde) [reply packages/babel-plugin/src/utils/css-builders.ts]: The condition for checking `ObjectProperty` is not required.

File: packages/babel-plugin/src/utils/css-builders.ts [hunk 7]
```diff
@@ -633,6 +643,31 @@ const extractTemplateLiteral = (node: t.TemplateLiteral, meta: Metadata): CSSOut
   const literalResult = node.quasis.reduce<string>((acc, quasi, index): string => {
     const nodeExpression = node.expressions[index] as t.Expression | undefined;
 
+    // Convert destructed props into structed props
+    // Babel was not able to differentiate between destrcuted prop and global variable if they have same name
+    if (
```
[the remark this answers is not in the document]
Comment (harshmalde) [reply packages/babel-plugin/src/utils/css-builders.ts:677]: Abstracted the code into another method

File: stories/css-shadow-variable-clash.tsx (hunk 1 above)
Comment (harshmalde) [reply stories/css-shadow-variable-clash.tsx]: Updated the component name and it's content

File: stories/css-shadow-variable-clash.tsx (hunk 2 above)
Comment (harshmalde) [reply stories/css-shadow-variable-clash.tsx:19]: Added another example without props

File: stories/css-shadow-variable-clash.tsx (hunk 4 above)
Comment (harshmalde) [reply stories/css-shadow-variable-clash.tsx]: Updated the component name

File: stories/css-shadow-variable-clash.tsx (hunk 3 above)
Comment (harshmalde) [reply stories/css-shadow-variable-clash.tsx]: Added `console.log()`

File: .changeset/slow-ducks-sneeze.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+'@compiled/babel-plugin': minor
```
Comment (pancaspe87) [inline .changeset/slow-ducks-sneeze.md:2]: I think this should be a patch

Please mention the fix in the description, eg "Fixed shadow variables clashing with destructured props"

**Review by pancaspe87 — CHANGES_REQUESTED** [review]
There is still a VR fail to address.

`FAIL chrome.docker/chrome.laptop/conditional rules/styled/Condition With Destructured Props Key Value String: Failed to render with error "primary is not defined"`

This is caused by a story within `stories/conditional-rules-styled.tsx`

The fix doesn't seem to take into account cases such as 
```
const DestructuredPropsKeyValueString = styled.div<TextProps>`
  ${({ isPrimary: primary }) => (primary ? 'color: green' : 'color: red')};
`;
```

File: packages/babel-plugin/src/utils/css-builders.ts [hunk 9]
```diff
@@ -619,6 +620,118 @@ const extractObjectExpression = (node: t.ObjectExpression, meta: Metadata): CSSO
...
+};
+
+/**
+ * Convert `consequent` and `alternate` of Conditional expressions into Member Expression
+ * @param node Expression which has to be converted into Member Expression
+ * @param parameters Deconstructored props
+ */
+const modifyTemplateLiterals = (node: t.Expression, parameters: FunctionParameters[]) => {
+  if (t.isTemplateLiteral(node)) {
+    let count = -1;
```
Comment (pancaspe87) [inline packages/babel-plugin/src/utils/css-builders.ts:682]: You can probably use `node.expressions[0]` directly instead of the loop.

**Review by pancaspe87 — DISMISSED** [review]

File: stories/css-shadow-variable-clash.tsx [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+export default {
+  title: 'css prop/variable clash',
+};
+
+const isPrimary = false;
+const Component = styled.div`
+  color: ${({ isPrimary }) => (isPrimary ? 'green' : 'red')};
+`;
+console.log(isPrimary);
```
Comment (JakeLane) [inline stories/css-shadow-variable-clash.tsx:11]: leftover debug code

**Review by JakeLane — CHANGES_REQUESTED** [review]
Looks good! Just the leftover console log

File: packages/babel-plugin/src/styled/__tests__/behaviour.test.ts [hunk 11]
```diff
@@ -1195,15 +1194,15 @@ describe('styled component behaviour', () => {
 
       const Component = styled.div\`
         :before {
-          content: '\${({ isOpen }) => isOpen ? 'show less' : 'show more'}';
+          content: \${({ isOpen }) => isOpen ? 'show less' : 'show more'};
```
[the remark this answers is not in the document]
Comment (harshmalde) [reply packages/babel-plugin/src/styled/__tests__/behaviour.test.ts]: updated the test

File: packages/babel-plugin/src/utils/css-builders.ts [hunk 12]
```diff
@@ -619,6 +620,118 @@ const extractObjectExpression = (node: t.ObjectExpression, meta: Metadata): CSSO
...
+  parameters: FunctionParameters[],
+  prevKey?: string
+): string => {
+  for (const param of parameters) {
+    if (Array.isArray(param.value)) {
+      prevKey = param.key + '.';
+      getPropsNotDestructured(deconstructedProp, param.value, prevKey);
+    }
+
+    const parameter = parameters.find((x) => x.value == deconstructedProp);
```
[the remark this answers is not in the document]
Comment (harshmalde) [reply packages/babel-plugin/src/utils/css-builders.ts:652]: added comments to describe use of alias

File: packages/babel-plugin/src/utils/css-builders.ts (hunk 9 above)
Comment (harshmalde) [reply packages/babel-plugin/src/utils/css-builders.ts]: removed for loop

File: stories/css-shadow-variable-clash.tsx (hunk 10 above)
Comment (pancaspe87) [reply stories/css-shadow-variable-clash.tsx]: I think that was the quickest workaround to use the variable, otherwise the build will fail.

Comment (harshmalde) [reply stories/css-shadow-variable-clash.tsx]: I have removed the console.log and all checks has passed.

**Review by JakeLane — APPROVED** [review]

**Review by pancaspe87 — APPROVED** [review]

**Pull request merged**